### PR TITLE
eos-diagnostics: Add hardware info dump mode

### PR DESCRIPTION
**Pull request by dsd**
We currently collect hardware info manually and populate wiki pages for each supported product.

With the increased number of 3rd party products, we need something more scalable and we need an easy way of receiving this info from 3rd parties.

Add a new mode to eos-diagnostics which only outputs information relevant to hardware enablement efforts, and output in a more organised format (1 file per information item, in a tar archive). This also organises the information a bit better in the standard mode and adds more information to the dumps.
https://phabricator.endlessm.com/T14154

### CODE REVIEW

File: eos-tech-support/eos-diagnostics [hunk 1]
```diff
@@ -383,15 +587,90 @@ function dumpDiagnostics(filename) {
...
+	print("ERROR: Failed to create archive");
+	return false;
+    }
+
+    print("Wrote " + filename);
+    return true;
+}
+
 let filename;
+if (ARGV[0] == "-h" || ARGV[0] == "--hardware") {
```
Comment (wjt) [inline eos-tech-support/eos-diagnostics]: Drive-by pet peeve bikeshedding: can I suggest `-H` rather than `-h`? I'd generally expect lowercase `-h` to be shorthand for `--help`.

File: eos-tech-support/eos-diagnostics [hunk 2]
```diff
@@ -196,180 +196,127 @@ function tryReadFile(filename, fallbackMessage) {
     }
 }
 
+let diagnostics = [
+    {
+	title: "EndlessOS version",
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Please use single quotes (multiple occurrences)

File: eos-tech-support/eos-diagnostics [hunk 3]
```diff
@@ -196,180 +196,127 @@ function tryReadFile(filename, fallbackMessage) {
...
-    fullDump += '\n';
-
-    fullDump += '==================\n';
-    fullDump += '= Journal output =\n';
-    fullDump += '==================\n';
-    fullDump += '\n';
-    fullDump += trySpawn('journalctl --no-pager');
+    for (var i = 0; i < diagnostics.length; i++) {
+        let diag = diagnostics[i];
+        fullDump += "=".repeat(diag['title'].length + 4) + '\n';
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Please use `diag.title` instead

File: eos-tech-support/eos-diagnostics [hunk 4]
```diff
@@ -196,180 +196,127 @@ function tryReadFile(filename, fallbackMessage) {
...
-    fullDump += '\n';
-    fullDump += trySpawn('systemd-analyze');
-    fullDump += '\n';
-
-    fullDump += '==================\n';
-    fullDump += '= Journal output =\n';
-    fullDump += '==================\n';
-    fullDump += '\n';
-    fullDump += trySpawn('journalctl --no-pager');
+    for (var i = 0; i < diagnostics.length; i++) {
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Prefer `let` to `var`, but you can also use `diagnostics.forEach()`

File: eos-tech-support/eos-diagnostics [hunk 5]
```diff
@@ -196,180 +196,127 @@ function tryReadFile(filename, fallbackMessage) {
...
+        let diag = diagnostics[i];
+        fullDump += "=".repeat(diag['title'].length + 4) + '\n';
+        fullDump += "= " + diag['title'] + ' =\n';
+        fullDump += "=".repeat(diag['title'].length + 4) + '\n';
+        let content = diag['content']().toString().trim();
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: The calls to `toString()` and `trim()` do not seem required here, since we did not have them before. Any reason you introduced them?

File: eos-tech-support/eos-diagnostics [hunk 6]
```diff
@@ -28,42 +28,30 @@ function collectGraphicsRenderer() {
...
-	    } else {
-	        output += 'not available\n';
-	    }
 	}
     });
 
+    try {
+	let [res, contents] = GLib.file_get_contents("/proc/device-tree/compatible");
+	output += 'compatible: ' + contents;
+
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Remove empty line

File: eos-tech-support/eos-diagnostics [hunk 7]
```diff
@@ -304,11 +542,14 @@ function dumpDiagnostics(filename) {
 
     for (var i = 0; i < diagnostics.length; i++) {
         let diag = diagnostics[i];
+        let content = diag['content']();
+        if (content === null || content == '')
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: `''` evaluates to false-y in JS, so you can just use `if (!content)` here

File: eos-tech-support/eos-diagnostics [hunk 8]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+    }
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+	let dev = info.get_name();
+	let devDir = dir.resolve_relative_path(dev).get_path();
+
+	let initPins = tryReadFile(devDir + '/init_pin_configs').toString();
+	let driverPins = tryReadFile(devDir + '/driver_pin_configs').toString();
+	if (initPins == '') {
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use `if (!initPins)`

File: eos-tech-support/eos-diagnostics [hunk 9]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+	    continue;
+	}
+
+	output += dev + ':\n';
+	output += 'vendor: ' + tryReadFile(devDir + '/vendor_id')
+	output += 'subsystem: ' + tryReadFile(devDir + '/subsystem_id')
+	output += 'init_pin_configs:\n' + initPins.trim() + '\n';
+	if (driverPins != '') {
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use `if (driverPins)`

File: eos-tech-support/eos-diagnostics [hunk 10]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+	    continue;
+	}
+	output += collectHdaProcCodecInfo(dir.resolve_relative_path(name));
+    }
+
+    return output;
+}
+
+function collectHdaSysInfo() {
+    let dir = Gio.File.new_for_path("/sys/class/sound");
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use single quotes

File: eos-tech-support/eos-diagnostics [hunk 11]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+	return '';
+    }
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+	let dev = info.get_name();
+	let devDir = dir.resolve_relative_path(dev).get_path();
+
+	let initPins = tryReadFile(devDir + '/init_pin_configs').toString();
+	let driverPins = tryReadFile(devDir + '/driver_pin_configs').toString();
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Are these not strings already?

File: eos-tech-support/eos-diagnostics [hunk 12]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+	fileEnum = dir.enumerate_children('standard::name',
+	                                  Gio.FileQueryInfoFlags.NONE, null);
+    } catch (e) {
+	return '';
+    }
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+	let dev = info.get_name();
+	let devDir = dir.resolve_relative_path(dev).get_path();
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: I think you can use `g_file_enumerator_get_child()` for this

File: eos-tech-support/eos-diagnostics [hunk 13]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+	return '';
+    }
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+	let name = info.get_name();
+	if (name == "cards" || !name.startsWith("card")) {
+	    continue;
+	}
+	output += collectHdaProcCodecInfo(dir.resolve_relative_path(name));
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use `g_file_enumerator_get_child()`

Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: (also other occurrences throughout)

File: eos-tech-support/eos-diagnostics (hunk 10 above)
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: (also other occurrences throughout)

File: eos-tech-support/eos-diagnostics [hunk 14]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+     * print useful info about both entry types. */
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+	let devDir = dir.resolve_relative_path(info.get_name()).get_path();
+	output += info.get_name() + '\n';
+
+	output += '  name: ' + tryReadFile(devDir + '/name');
+
+        let contents = tryReadFile(devDir + '/firmware_node/path');
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Indentation seems off

File: eos-tech-support/eos-diagnostics [hunk 15]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+    let output = '';
+
+    try {
+	fileEnum = dir.enumerate_children('standard::name',
+	                                  Gio.FileQueryInfoFlags.NONE, null);
+    } catch (e) {
+	return '';
+    }
+
+    /* This directory mixes i2c hosts and devices. Rather than try to
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Don't use C-style comments

Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: (other occurrences throughout)

File: eos-tech-support/eos-diagnostics [hunk 16]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+	let driver = GLib.file_read_link(path + '/device/driver');
+	output += prefix + ' driver: ' + GLib.path_get_basename(driver) + '\n';
+    } catch (e) { }
+
+    try {
+	let driver = GLib.file_read_link(path + '/driver');
+	output += prefix + ' driver: ' + GLib.path_get_basename(driver) + '\n';
+    } catch (e) { }
+
+
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Remove empty line

File: eos-tech-support/eos-diagnostics [hunk 17]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+    for (var i = 0; i < attrs.length; i++) {
+	let path = dir.resolve_relative_path(attrs[i]).get_path();
+        let contents = tryReadFile(path);
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Indentation looks a bit off

File: eos-tech-support/eos-diagnostics [hunk 18]
```diff
@@ -55,6 +55,217 @@ function collectBoardInfo() {
...
+    try {
+	fileEnum = dir.enumerate_children('standard::name,standard::type,standard::is-symlink',
+	                                  Gio.FileQueryInfoFlags.NONE, null);
+    } catch (e) {
+	return '';
+    }
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+        if (info.get_is_symlink() || info.get_attribute_uint32("standard::type") != 2)
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use `Gio.FileType.DIRECTORY`

File: eos-tech-support/eos-diagnostics [hunk 19]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env gjs
 
+const System = imports.system;
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Prefer alphabetical order

File: eos-tech-support/eos-diagnostics (hunk 1 above)
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Agreed with @wjt here

File: eos-tech-support/eos-diagnostics [hunk 20]
```diff
@@ -568,15 +584,90 @@ function dumpDiagnostics(filename) {
...
+
+    print("Wrote " + filename);
+    return true;
+}
+
 let filename;
+if (ARGV[0] == "-h" || ARGV[0] == "--hardware") {
+    filename = ARGV[1];
+
+    if (GLib.get_user_name() != "root") {
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: I think a more typical way of doing this check would be checking the user ID...
Perhaps you can check it by creating a `GCredentials` object?

File: eos-tech-support/eos-diagnostics [hunk 21]
```diff
@@ -568,15 +584,90 @@ function dumpDiagnostics(filename) {
...
+	print("ERROR: For hardware info, run eos-diagnostics as root.");
+	System.exit(1);
+    }
+
+    if (!filename) {
+	let date = GLib.DateTime.new_now_local();
+	filename = 'eos-hwinfo-' + date.format('%y%m%d_%H%M%S_UTC%z');
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Should this mode not default to the home directory, like the other mode?

File: eos-tech-support/eos-diagnostics [hunk 22]
```diff
@@ -568,15 +584,90 @@ function dumpDiagnostics(filename) {
...
+
+function dumpHardwareInfo(filename) {
+    let dir = "/tmp/" + filename;
+    GLib.mkdir_with_parents(dir, parseInt("0700", 8));
+
+    copyAcpiFirmware(dir);
+
+    for (var i = 0; i < diagnostics.length; i++) {
+        let diag = diagnostics[i];
+        if (!diag['hardwareInfo'])
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use `diag.hardwareInfo`

File: eos-tech-support/eos-diagnostics [hunk 23]
```diff
@@ -568,15 +584,90 @@ function dumpDiagnostics(filename) {
...
+            continue;
+
+        let name = diag['title'].toLowerCase().replace(' ', '_') + '.txt';
+
+        let content = diag['content']();
+        if (content === null || content == '')
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: See other similar comment about just using `!content`

File: eos-tech-support/eos-diagnostics (hunk 21 above)
Comment (dsd) [inline eos-tech-support/eos-diagnostics]: It would get saved in root's home directory. I think its best to make it save in the current directory so that you can just open a terminal and do "sudo eos-diagnostics" and then have access to the resultant file in the file browser.

File: eos-tech-support/eos-diagnostics (hunk 11 above)
Comment (dsd) [inline eos-tech-support/eos-diagnostics]: No, when the file doesn't exist, tryReadFile returns an object of type object. I don't understand why.

**Comment by dsd** [comment]
Figured out the toString thing and made all of the remaining changes, ready for review again.

**Comment by dsd** [comment]
hold on, I tested the wrong version before pushing, let me double check

**Comment by dsd** [comment]
Now tested properly, ready for review

File: eos-tech-support/eos-diagnostics [hunk 24]
```diff
@@ -28,42 +29,237 @@ function collectGraphicsRenderer() {
...
+
+    return output;
+}
+
+function collectMmcDevice(dir, level) {
+    let fileEnum;
+    let output = '';
+    let attrs = ['cid', 'manfid', 'oemid', 'name', 'prv', 'serial', 'date',
+	         'hwrev', 'fwrev', 'csd', 'scr', 'vendor', 'device'];
+    let prefix = ' '.repeat(level)
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Missing semicolon

File: eos-tech-support/eos-diagnostics [hunk 25]
```diff
@@ -28,42 +29,237 @@ function collectGraphicsRenderer() {
...
+    let path = dir.get_path()
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Missing semicolon

File: eos-tech-support/eos-diagnostics [hunk 26]
```diff
@@ -28,42 +29,237 @@ function collectGraphicsRenderer() {
...
+	output += prefix + ' driver: ' + GLib.path_get_basename(driver) + '\n';
+    } catch (e) { }
+
+    try {
+	let driver = GLib.file_read_link(path + '/driver');
+	output += prefix + ' driver: ' + GLib.path_get_basename(driver) + '\n';
+    } catch (e) { }
+
+    for (var i = 0; i < attrs.length; i++) {
+	let path = dir.resolve_relative_path(attrs[i]).get_path();
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: This shadows the variable with the same name at the top

File: eos-tech-support/eos-diagnostics [hunk 27]
```diff
@@ -28,42 +29,237 @@ function collectGraphicsRenderer() {
...
+	fileEnum = dir.enumerate_children('standard::name,standard::type,standard::is-symlink',
+	                                  Gio.FileQueryInfoFlags.NONE, null);
+    } catch (e) {
+	return '';
+    }
+
+    let info;
+    while ((info = fileEnum.next_file(null))) {
+	let fileType = info.get_attribute_uint32('standard::type');
+        if (info.get_is_symlink() || fileType != Gio.FileType.DIRECTORY)
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Indentation is a bit odd

File: eos-tech-support/eos-diagnostics [hunk 28]
```diff
@@ -383,15 +582,89 @@ function dumpDiagnostics(filename) {
     }
 }
 
-let fileArg = ARGV[0];
+function copyAcpiFirmware(output)
+{
+    let dir = Gio.File.new_for_path('/sys/firmware/acpi/tables');
+    let fileEnum;
+
+    output = Gio.File.new_for_path(output);
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Nitpick: use a different variable

File: eos-tech-support/eos-diagnostics [hunk 29]
```diff
@@ -383,15 +582,89 @@ function dumpDiagnostics(filename) {
...
+
+    for (var i = 0; i < diagnostics.length; i++) {
+        let diag = diagnostics[i];
+        if (!diag.hardwareInfo)
+            continue;
+
+        let name = diag.title.toLowerCase().replace(' ', '_') + '.txt';
+
+        let content = diag.content();
+        if (content === null || content == '')
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: Use `if (!content)`

File: eos-tech-support/eos-diagnostics [hunk 30]
```diff
@@ -383,15 +582,89 @@ function dumpDiagnostics(filename) {
...
+    return true;
+}
+
 let filename;
+if (ARGV[0] == '-H' || ARGV[0] == '--hardware') {
+    filename = ARGV[1];
 
-if (fileArg) {
-    filename = fileArg;
+    if (GLib.get_user_name() != 'root') {
```
Comment (cosimoc) [inline eos-tech-support/eos-diagnostics]: See my other comment about checking the user ID and not the name here.

**Comment by dsd** [comment]
Addressed those comments.
